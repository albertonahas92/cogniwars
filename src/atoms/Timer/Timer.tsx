import React, { FC, useEffect, useRef, useState } from "react"

export const Timer: FC<Props> = ({
  startingTime,
  active,
  onTimerStop,
  countdown,
  endtime = 0,
  onTimeChange,
}) => {
  const time = useRef(startingTime || 0)
  const [timeRef, setTimeRef] = useState(startingTime || 0)

  const endTime = useRef(endtime || 0)

  useEffect(() => {
    const timer = setInterval(() => {
      if (!active || (countdown && time.current >= endTime.current)) {
        clearInterval(timer)
        // Finish
        onTimerStop?.(time.current)
      } else {
        time.current += 1
        setTimeRef(time.current)
        onTimeChange?.(time.current)
      }
    }, 1000)

    return () => {
      clearInterval(timer)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [active])

  useEffect(() => {
    time.current = startingTime || 0
    setTimeRef(time.current)
    onTimeChange?.(time.current)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [startingTime])

  useEffect(() => {
    endTime.current = endtime || 0
  }, [endtime])

  const displayTime = () =>
    countdown
      ? new Date(Math.max(endtime - timeRef, 0) * 1000)
          .toISOString()
          .substr(14, 5)
      : new Date(timeRef * 1000).toISOString().substr(14, 5)

  return (
    <span aria-label="timer" className="timer">
      {displayTime()}
    </span>
  )
}

interface Props {
  startingTime?: number
  active: boolean
  countdown?: boolean
  endtime?: number
  onTimerStop?: (elapsedTime: number) => void
  onTimeChange?: (time: number) => void
}
