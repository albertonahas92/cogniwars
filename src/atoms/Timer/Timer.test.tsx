import React from "react"
import { mount, shallow } from "enzyme"
import { Timer } from "./Timer"
import waitFor from "../../utils/waitFor"

describe("Timer widget", function () {
  console.error = jest.fn()
  it("renders without crashing", () => {
    const component = shallow(<Timer active={true} />)
    expect(component).toMatchSnapshot()
  })

  it("starts at 00:00", () => {
    const component = shallow(<Timer active={true} />)
    expect(component.find(`[aria-label="timer"]`).text()).toContain("00:00")
  })

  it("starts with startingTime", () => {
    const component = shallow(<Timer startingTime={30} active={true} />)
    expect(component.find(`[aria-label="timer"]`).text()).toContain("00:30")
  })

  it("increases time and calls onTimeChange", async () => {
    const onTimeChangeMock = jest.fn((time) => time)
    const component = mount(
      <Timer onTimeChange={onTimeChangeMock} active={true} />
    )
    await waitFor(() => {
      component.update()
      expect(onTimeChangeMock).toHaveBeenCalledWith(1)
      expect(component.find(`[aria-label="timer"]`).text()).toContain("00:01")
    }, 2000)
  })

  it("should end with endTime and call onTimerStop with countdown", async () => {
    const onTimerStopMock = jest.fn((time) => time)
    const component = mount(
      <Timer
        endtime={1}
        startingTime={0}
        onTimerStop={onTimerStopMock}
        active={true}
        countdown={true}
      />
    )
    await waitFor(() => {
      component.update()
      expect(onTimerStopMock).toHaveBeenCalled()
    }, 4000)
  })
})
