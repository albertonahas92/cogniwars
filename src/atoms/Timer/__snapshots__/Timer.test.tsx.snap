// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer widget renders without crashing 1`] = `ShallowWrapper {}`;
